lib/Utils/AttributeUtils.cpp

```diff
@@ -97,6 +102,70 @@ FailureOr<Attribute> findAttributeAssociatedWith(Value value,
   return attr;
 }
 
+void setAttributeForBlockArgument(BlockArgument blockArg, StringRef attrName,
+                                  Attribute attr) {
+  auto *parentOp = blockArg.getOwner()->getParentOp();
+  assert(parentOp != nullptr &&
+         "Missing parent op! Was this value not properly remapped?");
+
+  llvm::TypeSwitch<Operation *>(parentOp)
+      .Case<FunctionOpInterface>([&](auto op) {
+        return op.setArgAttr(blockArg.getArgNumber(), attrName, attr);
+      })
+      .Case<affine::AffineForOp>([&](affine::AffineForOp op) {
+        // For op has as its first block argument the induction
+        // variable, which does not correspond to a single operand.
+        if (blockArg.getArgNumber() == 0) return;
+
+        auto argAttrInterface = cast<OperandAndResultAttrInterface>(*op);
+        auto initArg = op.getInits()[blockArg.getArgNumber() - 1];
+        int operandNumber = getOperandNumber(op, initArg);
+        if (operandNumber == -1) return;
+        return argAttrInterface.setOperandAttr(operandNumber, attrName, attr);
+      })
+      .Case<OperandAndResultAttrInterface>([&](auto op) {
+        return op.setOperandAttr(blockArg.getArgNumber(), attrName, attr);
+      });
+}
+
+void setAttributeForResult(Value result, StringRef attrName, Attribute attr) {
+  auto *parentOp = result.getDefiningOp();
+  assert(parentOp &&
+         "Missing defining op, but the input value is not a BlockArgument. "
+         "This is madness.");
+  int resultNumber = cast<OpResult>(result).getResultNumber();
+
+  llvm::TypeSwitch<Operation *>(parentOp)
+      .Case<OperandAndResultAttrInterface>(
+          [&](auto op) { op.setResultAttr(resultNumber, attrName, attr); })
+      .Default([&](Operation *op) {
+        // If the op has an existing array attr, set the attribute in the
+        // existing array attr's index.
+        auto existingAttr = op->getAttr(attrName);
+        if (auto existingArrayAttr =
+                dyn_cast_or_null<ArrayAttr>(existingAttr)) {
```

Comment:
Still ArrayAttr here is quite ambiguous. For example `secret.execution_result` is an ArrayAttr and it does not have the semantic of each element corresponding to each element. If we setAttributeForResult for it twice, the second one will run into assertion.

I would suggest determining the shape (ArrayAttr of ArrayAttr, or ArrayAttr of Attr, or Attr) by the number of result.

Anyway, lets merge it and when we run into issue other PR can change it.

Reply:
I don't think every attribute necessarily needs to use these utils, it's mainly meant for a common case.